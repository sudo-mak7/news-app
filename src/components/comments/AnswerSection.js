import React from 'react'
import { Comment } from 'semantic-ui-react'
import CommentComponent from './CommentComponent'

const AnswerSection = ({ isCollapsed }) => {
  const commentOne = {
    author: `Elliot Fu`,
    text: `No, it wont!`,
    hasAnswer: false
  }
  const commentTwo = {
    author: `Jenny Hess`,
    text: `Maybe it would.`,
    hasAnswer: false
  }

  return (
    <Comment.Group collapsed={ isCollapsed }>
      <CommentComponent
        author={ commentOne.author }
        text={ commentOne.text }
        hasAnswer={ commentOne.hasAnswer }
      />
      <CommentComponent
        author={ commentTwo.author }
        text={ commentTwo.text }
        hasAnswer={ commentTwo.hasAnswer }
      />
    </Comment.Group>
  )
}

export default AnswerSection