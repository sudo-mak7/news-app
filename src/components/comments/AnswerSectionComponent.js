import React from 'react'
import { Comment, Loader } from 'semantic-ui-react'
import CommentComponent from './CommentComponent'

const AnswerSectionComponent = ({ isCollapsed, answers, answersLoading }) => {
  const commentsRender = answers.map(a =>
    <CommentComponent
      key={ a.id }
      id={ a.id }
      { ...a }
    />
  )

  return (
    <Comment.Group collapsed={ isCollapsed }>
      { answersLoading
          ? <Loader active inline='centered'/>
          : commentsRender
      }
    </Comment.Group>
  )
}

export default AnswerSectionComponent