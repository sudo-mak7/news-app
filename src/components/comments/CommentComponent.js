import React, { useState } from 'react'
import { Label, Comment, Header } from 'semantic-ui-react'
import AnswerSectionComponent from './AnswerSectionComponent'
import { getAnswers, getCurrentItemById } from '../../api/api'

const CommentComponent = ({ id, by, text, kids, dead, deleted }) => {

  const [ answers, setAnswers ] = useState([])
  const [ answersLoading, setAnswersLoading ] = useState([])
  const [ error, setError ] = useState('')

  const errorMessage = 'Error loading comments :('

  const [ isCollapsed, setIsCollapsed ] = useState(true)

  const handleCollapse = (e) => {
    isCollapsed === false
      ? setIsCollapsed(true)
      : setIsCollapsed(false)

    const id = e.currentTarget.id

    setAnswersLoading(true)

    getCurrentItemById(id)
      .then(data => {
        !data
          ? setError(errorMessage)
          : getAnswers(
            data.kids
              ? data.kids
              : setError(errorMessage)
            )
              .then(data => {
                !data
                  ? setError(errorMessage)
                  : setAnswers(data)
                setAnswersLoading(false)
              })
    })
  }

  return (
    <Comment id={ id }>
      <Comment.Avatar
      as='a'
      src='https://transcount.com/views/images/features/avatar.png'
    />

    { dead || deleted
      ? <Comment.Content>
        <Comment.Author as='a' style={{ color: 'grey' }}>DELETED</Comment.Author>
        <Comment.Text style={{ color: 'grey' }}>Comment deleted</Comment.Text>

        <Comment.Actions>
          { !kids
              ? ''
              : <Label
                size='mini'
                style={{ cursor: 'pointer' }}
                id={ id }
                onClick={ (e) => handleCollapse(e) }
              >
                { isCollapsed === true
                    ? 'Show answers'
                    : 'Hide answers'
                }
              </Label>
          }
        </Comment.Actions>
      </Comment.Content>

      : <Comment.Content>
        <Comment.Author as='a'>{ by }</Comment.Author>
        <Comment.Text
          dangerouslySetInnerHTML={{ __html: text }}
        />
        <Comment.Actions>
          { !kids
              ? ''
              : <Label
                size='mini'
                style={{ cursor: 'pointer' }}
                id={ id }
                onClick={ (e) => handleCollapse(e) }
              >
                { isCollapsed === true
                    ? 'Show answers'
                    : 'Hide answers'
                }
              </Label>
          }
        </Comment.Actions>
      </Comment.Content>
    }

    { !kids
        ? ''
        : error
            ? <Header as='h4' textAlign='center'>{ error }</Header>
            : <AnswerSectionComponent
                isCollapsed={ isCollapsed }
                answers={ answers }
                answersLoading={ answersLoading }
              />
    }
  </Comment>
  )
}

export default CommentComponent