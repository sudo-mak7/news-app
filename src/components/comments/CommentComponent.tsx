import * as React from 'react'
import { useState } from 'react'
import DOMPurify from 'dompurify'
import { Label, Comment, Header } from 'semantic-ui-react'
import { useAppDispatch } from '@redux/reduxHooks'
import { useAppSelector } from '@redux/reduxHooks'
import { getCommentsErrorSelector } from '@redux/selectors'
import { fetchAnswers } from '@redux/comments/answersSlice'
import { paths } from '@api/url'
import AnswerSectionComponent from '@components/comments/AnswerSectionComponent'
import { CommentsInterface } from '@common/types/commentsInterface'
import { ERROR_LOADING_COMMENTS_MESSAGE } from '@common/messages/errors'

const CommentComponent = (
  { id, by, text, kids, dead, deleted }: CommentsInterface
): JSX.Element => {
  const commentsLoadingError = useAppSelector(getCommentsErrorSelector)

  const [ isCollapsed, setIsCollapsed ] = useState<boolean>(true)
  const [ answers, setAnswersState ] = useState<CommentsInterface[]>([])
  const [ isLoading, setIsLoading ] = useState<boolean>(false)
  const [ error, setError ] = useState<any>(commentsLoadingError)

  const dispatch = useAppDispatch()

  const handleCollapse = (e: React.MouseEvent<HTMLButtonElement>): void => {
    setIsCollapsed(prev=>!prev)

    const id = Number(e.currentTarget.id)

    if (isCollapsed) {
      dispatch(fetchAnswers({ id, setAnswersState, setIsLoading, setError }))
    }
  }

  const sanitizedHTMLText = () => ({
    __html: DOMPurify.sanitize(text)
  })

  const errorRender =
    <Header
      as='h4'
      textAlign='center'>
      { ERROR_LOADING_COMMENTS_MESSAGE }
    </Header>

  const commentActionsRender =
    <Comment.Actions>
      { kids
          ? <Label
              size='mini'
              style={{ cursor: 'pointer' }}
              id={ id }
              onClick={
                (e: React.MouseEvent<HTMLButtonElement>): void => handleCollapse(e)
              }
            >
              { isCollapsed ? 'Show answers' : 'Hide answers' }
            </Label>
          : ''
      }
    </Comment.Actions>

  const commentContentRender =
    dead || deleted
      ? <Comment.Content>
          <Comment.Author as='a' style={{ color: 'grey' }}>DELETED</Comment.Author>
          <Comment.Text style={{ color: 'grey' }}>Comment deleted</Comment.Text>
          { commentActionsRender }
        </Comment.Content>

      : <Comment.Content>
          <Comment.Author as='a'>{ by }</Comment.Author>
          <Comment.Text
            dangerouslySetInnerHTML={ sanitizedHTMLText() }
          />
          { commentActionsRender }
        </Comment.Content>

  const answersSectionRender =
    kids
      ? <AnswerSectionComponent
          isCollapsed={ isCollapsed }
          answers={ answers }
          isLoading={ isLoading }
        />
      : ''

  const commentRender =
    <Comment id={ id }>
      <Comment.Avatar
        as='a'
        src={ paths.avatarUrl() }
      />
      { commentContentRender }
      { answersSectionRender }
    </Comment>

  return (
    error
      ? errorRender
      : commentRender
  )
}

export default CommentComponent