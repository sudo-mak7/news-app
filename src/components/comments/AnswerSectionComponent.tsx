import * as React from 'react'
import { Comment, Header, Loader } from 'semantic-ui-react'
import { useAppSelector } from '@redux/reduxHooks'
import { getCommentsErrorSelector } from '@redux/selectors'
import CommentComponent from '@components/comments/CommentComponent'
import { CommentsInterface } from '@common-types/commentsInterface'

interface AnswerSectionComponentInterface {
  isCollapsed: boolean,
  answers: CommentsInterface[],
  isLoading: boolean
}

const AnswerSectionComponent = (
  { isCollapsed, answers, isLoading }: AnswerSectionComponentInterface
): JSX.Element => {
  const commentsLoadingError = useAppSelector(getCommentsErrorSelector)

  const errorMessage = 'Error loading answers :('

  const commentsRender = answers.map((a: CommentsInterface) =>
    <CommentComponent
      key={ a.id }
      id={ a.id }
      { ...a }
    />
  )

  return (
    isCollapsed
      ? null
      : <Comment.Group>
          { isLoading
              ? <Loader active inline='centered'/>
              : commentsLoadingError
                  ? <Header as='h4' textAlign='center'>{ errorMessage }</Header>
                  : commentsRender
          }
        </Comment.Group>
  )
}

export default AnswerSectionComponent