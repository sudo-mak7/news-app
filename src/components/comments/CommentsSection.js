import React from 'react'
import { Comment } from 'semantic-ui-react'
import CommentComponent from './CommentComponent'

const CommentsSection = ({ comment })  => {
  return (
    <Comment.Group>
      <CommentComponent
        author={ comment.author }
        text={ comment.text }
        hasAnswer={ comment.hasAnswer }
      />
    </Comment.Group>
  )
}

export default CommentsSection