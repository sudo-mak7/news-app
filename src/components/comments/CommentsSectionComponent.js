import React, { useEffect } from 'react'
import { Comment, Header, Loader } from 'semantic-ui-react'
import CommentComponent from './CommentComponent'
import { useDispatch, useSelector } from 'react-redux'
import { getComments } from '../../api/api'
import { loading } from '../../redux/loading/loadingSlice'
import { setComments } from '../../redux/comments/commentsSlice'

const CommentsSectionComponent = ({ error, setError })  => {
  const news = useSelector(state => state.currentNews.currentNews)
  const isLoading = useSelector(state => state.loading.isLoading)
  const comments = useSelector(state => state.setComments.comments)

  const dispatch = useDispatch()

  const errorMessage = 'Error loading comments :('

  useEffect(() => {
    if (news?.hasOwnProperty('kids')) {
      dispatch(loading(true))

      getComments(
        news.kids
          ? news.kids
          : setError(errorMessage)
      ).then(data => {
        !data
          ? setError(errorMessage)
          : dispatch(setComments(data))
        dispatch(loading(false))
      })
    }
  }, [])

  const commentsRender = comments.map(c =>
      <CommentComponent
        key={ c?.id }
        id={ c?.id }
        { ...c }
      />
  )

  return (
    <Comment.Group>
      { isLoading
          ? <div style={{ height: '75px' }}>
              <Loader active/>
            </div>
          : ( error
                ? <Header as='h4' textAlign='center'>{ error }</Header>
                : commentsRender
          )
      }
    </Comment.Group>
  )
}

export default CommentsSectionComponent