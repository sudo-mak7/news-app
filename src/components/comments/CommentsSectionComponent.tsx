import * as React from 'react'
import { Comment, Loader } from 'semantic-ui-react'
import { useAppSelector } from '@redux/reduxHooks'
import {
  getCommentsLoaderSelector,
  getCommentsSelector
} from '@redux/selectors'
import CommentComponent from '@components/comments/CommentComponent'
import { CommentsInterface } from '@common/types/commentsInterface'

const CommentsSectionComponent = (): JSX.Element => {
  const isLoading = useAppSelector(getCommentsLoaderSelector)
  const comments = useAppSelector(getCommentsSelector)

  const commentsRender = comments.map((c: CommentsInterface) =>
    <CommentComponent
      key={ c?.id }
      id={ c?.id }
      { ...c }
    />
  )

  return (
    <Comment.Group>
      { isLoading
          ? <div style={{ height: '75px' }}>
              <Loader active/>
            </div>
          : commentsRender
      }
    </Comment.Group>
  )
}

export default CommentsSectionComponent