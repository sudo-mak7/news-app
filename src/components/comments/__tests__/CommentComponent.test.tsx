import * as React from 'react'
import * as reduxHooks from '@redux/reduxHooks'
import { store } from '@redux/store'
import { Provider } from 'react-redux'
import { render, screen, fireEvent } from '@testing-library/react'
import { useState } from 'react'
import CommentComponent from '@components/comments/CommentComponent'
import { commentsSlice, fetchComments } from '@redux/comments/commentsSlice'
import { ERROR_LOADING_COMMENTS_MESSAGE } from '@common/messages/errors'
import { CommentsInterface } from '@common/types/commentsInterface'
import { CommentsStateInterface } from '@redux/types'

const kids = [ 2922097, 2922429, 2924562, 2922709, 2922573, 2922140, 2922141 ]
const id = 2921983
const answer = {
  "by" : "pchristensen",
  "id" : 2922097,
  "parent" : 2921983,
  "text" : "Some text",
  "time" : 1314213033,
  "type" : "comment"
}

const initialState = {
  comments: [],
  loading: false,
  error: null
} as CommentsStateInterface

const commentsReducer = commentsSlice.reducer

const mockedUseState = jest.spyOn(React, 'useState')

const commentComponent = (
  { id, by, text, kids, dead, deleted }: CommentsInterface
): JSX.Element =>
  <Provider store={ store }>
    <CommentComponent
      id={ id }
      by={ by }
      text={ text }
      kids={ kids }
      dead={ dead }
      deleted={ deleted }
    />
  </Provider>

const mockedDispatch = jest.spyOn(reduxHooks, 'useAppDispatch')

describe('CommentComponent test', () => {
  test('CommentComponent should to match snapshot', () => {
    const props = {
      id: id,
      by: 'author',
      text: 'Some text',
      kids: [],
      dead: false,
      deleted: false
    }

    mockedUseState.mockImplementationOnce(() => [true, () => null]) // isCollapsed
    mockedUseState.mockImplementationOnce(() => [[], () => null]) // answers
    mockedUseState.mockImplementationOnce(() => [false, () => null]) // isLoading
    mockedUseState.mockImplementationOnce(() => [true, () => null]) // error

    render(commentComponent(props))
    expect(commentComponent(props)).toMatchSnapshot()
  })

  test('renders CommentComponent', () => {
    const props = {
      id: id,
      by: 'author',
      text: 'Some text',
      kids: [],
      dead: false,
      deleted: false
    }

    mockedUseState.mockImplementationOnce(() => [true, () => null]) // isCollapsed
    mockedUseState.mockImplementationOnce(() => [[], () => null]) // answers
    mockedUseState.mockImplementationOnce(() => [false, () => null]) // isLoading
    mockedUseState.mockImplementationOnce(() => [false, () => null]) // error

    render(commentComponent(props))

    const commentInCommentComponent = screen.getByTestId('comment')
    expect(commentInCommentComponent).toBeInTheDocument()
  })

  test('renders CommentComponent with answers', () => {
    const props = {
      id: id,
      by: 'author',
      text: 'Some text',
      kids: kids,
      dead: false,
      deleted: false
    }

    mockedUseState.mockImplementationOnce(() => [true, () => null]) // isCollapsed
    mockedUseState.mockImplementationOnce(() => [[ answer ], () => null]) // answers
    mockedUseState.mockImplementationOnce(() => [false, () => null]) // isLoading
    mockedUseState.mockImplementationOnce(() => [false, () => null]) // error

    render(commentComponent(props))

    const showAnswersButtonInCommentComponent = screen.getByText('Show answers')
    expect(showAnswersButtonInCommentComponent).toBeInTheDocument()
  })

  test('renders CommentComponent with no answers', () => {
    const props = {
      id: id,
      by: 'author',
      text: 'Some text',
      kids: null,
      dead: false,
      deleted: false
    }

    mockedUseState.mockImplementationOnce(() => [true, () => null]) // isCollapsed
    mockedUseState.mockImplementationOnce(() => [[], () => null]) // answers
    mockedUseState.mockImplementationOnce(() => [false, () => null]) // isLoading
    mockedUseState.mockImplementationOnce(() => [false, () => null]) // error

    render(commentComponent(props))

    const showAnswersButtonInCommentComponent = screen.queryByText('Show answers')
    expect(showAnswersButtonInCommentComponent).not.toBeInTheDocument()
  })

  test('renders CommentComponent when dead/deleted comment', () => {
    const props = {
      id: id,
      by: 'author',
      text: 'Some text',
      kids: [ null ],
      dead: true,
      deleted: true
    }

    mockedUseState.mockImplementationOnce(() => [true, () => null]) // isCollapsed
    mockedUseState.mockImplementationOnce(() => [[], () => null]) // answers
    mockedUseState.mockImplementationOnce(() => [false, () => null]) // isLoading
    mockedUseState.mockImplementationOnce(() => [false, () => null]) // error

    render(commentComponent(props))

    const deletedCommentInCommentComponent = screen.getByText(/comment deleted/i)
    expect(deletedCommentInCommentComponent).toBeInTheDocument()
  })

  test('button "Show answers" should changed to "Hide answers"', () => {
    const props = {
      id: id,
      by: 'author',
      text: 'Some text',
      kids: [],
      dead: false,
      deleted: false
    }

    mockedUseState.mockImplementationOnce(() => [false, () => null]) // isCollapsed
    mockedUseState.mockImplementationOnce(() => [[], () => null]) // answers
    mockedUseState.mockImplementationOnce(() => [false, () => null]) // isLoading
    mockedUseState.mockImplementationOnce(() => [false, () => null]) // error

    const dispatch = jest.fn()
    mockedDispatch.mockReturnValue(dispatch)

    render(commentComponent(props))
    const labelInCommentComponent = screen.getByTestId('button-label')

    fireEvent.click(labelInCommentComponent)

    const deletedCommentInCommentComponent = screen.getByText(/hide answers/i)
    expect(deletedCommentInCommentComponent).toBeInTheDocument()
  })

  test('should render error message', () => {
    mockedUseState.mockImplementationOnce(() => [true, () => null]) // isCollapsed
    mockedUseState.mockImplementationOnce(() => [[], () => null]) // answers
    mockedUseState.mockImplementationOnce(() => [false, () => null]) // isLoading
    mockedUseState.mockImplementationOnce(() => [true, () => null]) // error

    const state = commentsReducer(initialState, fetchComments.rejected)
    expect(state.error).not.toBeNull()

    render(
      <Provider store={ store }>
        <CommentComponent />
      </Provider>
    )

    const errorMessageInCommentComponent = screen.getByText(ERROR_LOADING_COMMENTS_MESSAGE)
    expect(errorMessageInCommentComponent).toBeInTheDocument()
  })

  test('dispatch should called 1 times by click', () => {
    const props = {
      id: id,
      by: 'author',
      text: 'Some text',
      kids: [123456],
      dead: false,
      deleted: false
    }

    mockedUseState.mockImplementationOnce(() => [true, () => null]) // isCollapsed
    mockedUseState.mockImplementationOnce(() => [[], () => null]) // answers
    mockedUseState.mockImplementationOnce(() => [false, () => null]) // isLoading
    mockedUseState.mockImplementationOnce(() => [false, () => null]) // error

    const dispatch = jest.fn()
    mockedDispatch.mockReturnValue(dispatch)

    render(commentComponent(props))
    const labelInCommentComponent = screen.getByTestId('button-label')

    fireEvent.click(labelInCommentComponent)
    expect(dispatch).toHaveBeenCalledTimes(1)
  })

  afterAll(() => {
    jest.clearAllMocks()
  })
})