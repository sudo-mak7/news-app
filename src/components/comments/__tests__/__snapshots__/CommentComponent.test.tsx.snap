// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentComponent test CommentComponent should to match snapshot 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <CommentComponent
    by="author"
    dead={false}
    deleted={false}
    id={2921983}
    kids={Array []}
    text="Some text"
  />
</Provider>
`;
